### ISSUE TITLE
IPeriodList does not include TzId during serialization

### ISSUE THREAD

**Issue by rianjs** [ical-org/ical.net#140]
``` csharp
[Test]
public void NoTzIdInPeriodListSerialization()
{
    var somePeriod = new Period(new CalDateTime(_nowTime, _tzid), new CalDateTime(_later, _tzid));
    var periodList = new PeriodList {somePeriod};
    periodList.TzId = _tzid;
    var someEvent = new Event
    {
        DtStart = new CalDateTime(_nowTime, _tzid),
        DtEnd = new CalDateTime(_later, _tzid),
        ExceptionDates = new List<IPeriodList> {periodList}
    };

    var calendar = new Calendar();
    calendar.Events.Add(someEvent);
    var serialized = SerializeToString(calendar);
    Assert.IsTrue(serialized.Contains("EXDATE;TZID=Eastern Standard Time:"));
}
```

Output is:

```
BEGIN:VCALENDAR
PRODID:-//github.com/rianjs/ical.net//NONSGML ical.net 2.2//EN
VERSION:2.0
BEGIN:VEVENT
DTEND;TZID=America/New_York:20160912T183326
DTSTAMP:20160912T213326Z
DTSTART;TZID=America/New_York:20160912T173326
EXDATE:20160912T173326/PT1H
SEQUENCE:0
UID:e2944545-37fe-47d9-bb74-19c134128023
END:VEVENT
END:VCALENDAR
```

EXDATE really should be `EXDATE;TZID=Eastern Standard Time:20160912T173326` (I think?)

**Comment by lngr** [comment]
After a long debugging session, we were stumbling over this issue.

If the server does not run in the same TZ as the event, effectively the EXDATEs are ignored, e.g., assume the server runs in UTC and we are using TZ of Europe/Berlin (UTC+0100):

This does not work (does NOT exclude the event at 2017-01-24):
```
BEGIN:VEVENT
DTEND;TZID=Europe/Berlin:20170117T171000
DTSTART;TZID=Europe/Berlin:20170117T164000
EXDATE:20170124T164000
RRULE:FREQ=WEEKLY
SEQUENCE:0
UID:45878d4d-ebe7-e72b-9ffd-22a7ac617444
END:VEVENT
```

This does work when deserializing (with 2.2.28) (note the different EXDATE definition).
```
BEGIN:VEVENT
DTEND;TZID=Europe/Berlin:20170117T171000
DTSTART;TZID=Europe/Berlin:20170117T164000
EXDATE;TZID=Europe/Berlin:20170124T164000
RRULE:FREQ=WEEKLY
SEQUENCE:0
UID:45878d4d-ebe7-e72b-9ffd-22a7ac617444
END:VEVENT
```

**Comment by rianjs** [comment]
Yeah, I think time zones are an area of weakness for ical.net, but they're more important that ever. When dday.ical was written (which is what ical.net has evolved from), it was the mid 2000s, and client-server scenarios weren't as prevalent as they are now.

Now everything's a service (or a client of a service), and these shortcomings are going to become more painful.

**Comment by aluxnimm** [comment]
Yes the timezone issues are our showstopper to migrate from dday.ical for outlookcaldavsynchronizer.
The code is painful slow compared to ical.net there but at least the VTIMEZONE serializer works more or less.

**Comment by rianjs** [comment]
I would love a PR that fixes it. :)

I improve ical.net as part of my day job--that's really the only way improvements get made--and I'm pretty liberal in who I give commit rights to. So open a reasonable PR, and I'll leave feedback, and then probably just give you commit rights.

**Comment by aluxnimm** [comment]
I will if time permits, but since the vcard library we use has even more problems and is abandoned as well, I started forking and improving that first :)

**Comment by rianjs** [comment]
I completely understand. :)

I recently added a "get occurrence" endpoint to a REST service here at work. No one is using it yet, but as soon as they start, this will become a higher priority I'm sure. :)

**Comment by rianjs** [comment]
@aluxnimm - nuget version 2.2.34 should include the fix for PeriodLists.

Note that you should explicitly set the `TzId` on the `PeriodList` itself. The serialization is a little backward in that while `CalDateTime`'s do indeed have a time zone, there are no correctness guarantees regarding the time zones of child `CalDateTime`s in your `PeriodList`. You'll have to enforce this in the application layer.

Example:

```csharp
[Test]
public void ExDateTimeZone_Tests()
{
    const string tzid = "Europe/Stockholm";

    //Repeat daily for 10 days
    var rrule = new RecurrencePattern(FrequencyType.Daily, 1) { Count = 10 };

    var e = new CalendarEvent
    {
        DtStart = new CalDateTime(_now, tzid),
        DtEnd = new CalDateTime(_later, tzid),
        RecurrenceRules = new List<RecurrencePattern> { rrule },
    };

    var exceptionDateList = new PeriodList { TzId = tzid };
    exceptionDateList.Add(new Period(new CalDateTime(_now.AddDays(1))));
    e.ExceptionDates.Add(exceptionDateList);

    var serialized = SerializeToString(e);
    const string expected = "TZID=Europe/Stockholm";
    Assert.AreEqual(3, Regex.Matches(serialized, expected).Count);

    e.ExceptionDates.First().Add(new Period(new CalDateTime(_now.AddDays(2))));
    serialized = SerializeToString(e);
    Assert.AreEqual(3, Regex.Matches(serialized, expected).Count);
}
```

In a perfect world, you could do something like:

```csharp
var time = DateTime.Parse("2017-04-06T14:09:55");
foo.Events.RecurrenceDates.Add(new Period(new CalDateTime(time, "America/New_York")))`;
foo.Events.RecurrenceDates.Add(new Period(new CalDateTime(time, "Europe/Stockholm")));
```

...and the library would create two separate `PeriodList`s, and make sure that each time zone is kept separately. But we're not there yet.

**Comment by aluxnimm** [comment]
Thanks for the update and great work! Hopefully, we can migrate soon. Any news on the VTIMEZONE serializing by chance?

**Comment by rianjs** [comment]
No, we don't use it at work, so I'm unlikely to get to it within a reasonable timeframe. (My main non-work priority is getting the .NET Core version working.)

**Comment by rianjs** [comment]
Fixed in nuget version 4.0.0: https://www.nuget.org/packages/Ical.Net/4.0.0
